File: librtEdge/CLI/src/Chain.cpp
/******************************************************************************
*
*  Chain.cpp
*
*  REVISION HISTORY:
*      7 JAN 2015 jcs  Created
*
*  (c) 1994-2015 Gatea, Ltd.
******************************************************************************/
#include "StdAfx.h"
#include <Chain.h>


namespace librtEdgePRIVATE
{

////////////////////////////////////////////////
//
//       c l a s s     C h a i n C
//
////////////////////////////////////////////////

//////////////////////////
// Constructor
//////////////////////////
ChainC::ChainC( IChain     ^cli,
                const char *svc,
                const char *link1 ) :
   _cli( cli ),
   _upd( gcnew librtEdge::rtEdgeData() ),
   RTEDGE::Chain( svc, link1 )
{
}

ChainC::~ChainC()
{
   _upd = nullptr;
}


//////////////////////////
// Asynchronous Callbacks
//////////////////////////
void ChainC::OnChainLink( const char *name, int nLnk, RTEDGE::Message &msg )
{
   _upd->Set( msg );
   _cli->OnLink( gcnew String( name ), nLnk, _upd );
}

void ChainC::OnChainData( const char      *name, 
                          int              pos, 
                          int              nUpd, 
                          RTEDGE::Message &msg )
{
   _upd->Set( msg );
   _cli->OnData( gcnew String( name ), pos, nUpd, _upd );
}

} // namespace librtEdgePRIVATE


namespace librtEdge
{

////////////////////////////////////////////////
//
//         c l a s s   C h a i n
//
////////////////////////////////////////////////

/////////////////////////////////
// Constructor / Destructor
/////////////////////////////////
Chain::Chain( String ^svc, String ^tkr ) :
   _chn( new librtEdgePRIVATE::ChainC( this, 
                                        rtEdge::_pStr( svc ),
                                        rtEdge::_pStr( tkr ) ) )
{
}

Chain::~Chain()
{
   delete _chn;
   _chn = (librtEdgePRIVATE::ChainC *)0;
}


/////////////////////////////////
// Access
/////////////////////////////////
RTEDGE::Chain &Chain::chn()
{
   return *_chn;
}

String ^Chain::svc()
{
   return gcnew String( chn().svc() );
}

String ^Chain::name()
{
   return gcnew String( chn().name() );
}

} // namespace librtEdge

File: librtEdge/CLI/examples/Chain/Chain.cs
/******************************************************************************
*
*  ChainTest.cs
*     Chain subscriber
*
*  REVISION HISTORY:
*     14 DEC 2014 jcs  Created.
*
*  (c) 1994-2014 Gatea, Ltd.
******************************************************************************/
using System;
using librtEdge;


class MyChain : Chain
{
   ////////////////////////////////
   // Constructor
   ////////////////////////////////
   public MyChain( string svc, string tkr ) :
      base( svc, tkr )
   { ; }


   ////////////////////////////////
   // Chain Interface
   ////////////////////////////////
   public override void OnLink( string name, int nLnk, rtEdgeData data )
   {
      Console.WriteLine( "OnLink( {0}, {1} )", nLnk, name );
   }

   public override void OnData( string name, int pos, int nUpd, rtEdgeData data )
   {
      string mt = data.MsgType();

      Console.WriteLine( "{0}( {1}, {2} ) : nUpd={3}", mt, pos, name, nUpd );
   }

}

class ChainTest : rtEdgeSubscriber
{
   ////////////////////////////////
   // Constructor
   ////////////////////////////////
   public ChainTest( string pSvr, string pUsr ) :
      base( pSvr, pUsr )
   {
   }

   ////////////////////////////////
   // rtEdgeSubscriber Interface
   ////////////////////////////////
   public override void OnConnect( string pConn, rtEdgeState state )
   {
      bool   bUp = ( state == rtEdgeState.edg_up );
      string pUp = bUp ? "UP" : "DOWN";

      Console.WriteLine( "CONN {0} : {1}", pConn, pUp );
   }

   public override void OnService( string pSvc, rtEdgeState state )
   {
      bool   bUp = ( state == rtEdgeState.edg_up );
      string pUp = bUp ? "UP" : "DOWN";

      if ( !pSvc.Equals( "__GLOBAL__" ) )
         Console.WriteLine( "SVC  {0} {1}", pSvc, pUp );
   }


   ////////////////////////////////
   // main()
   ////////////////////////////////
   public static int Main(String[] args)
   {
      try {
         ChainTest sub;
         MyChain   chn;
         int       i, argc;
         string    pEdg, pSvc, pTkr, pUsr;

         // [ <host:port> <Svc> <Tkr> <User> ]

         argc  = args.Length;
         pEdg  = "localhost:9998";
         pSvc  = "B-PIPE";
         pTkr  = "0#INDEX";
         pUsr  = "ChainTest";
         for ( i=0; i<argc; i++ ) {
            switch( i ) {
               case 0: pEdg = args[i]; break;
               case 1: pSvc = args[i]; break;
               case 2: pTkr = args[i]; break;
               case 3: pUsr = args[i]; break;
            }
         }
         Console.WriteLine( rtEdge.Version() );
         sub = new ChainTest( pEdg, pUsr );
         chn = new MyChain( pSvc, pTkr );
         Console.WriteLine( sub.Start() );
         sub.Subscribe( chn );
         Console.WriteLine( "Hit <ENTER> to terminate..." );
         Console.ReadLine();
         sub = null; // sub.Destroy();
      }
      catch( Exception e ) {
         Console.WriteLine( "Exception: " + e.Message );
      }
      return 0;
    }
}
